flatfs.go

```diff
@@ -739,60 +810,110 @@ func (fs *Datastore) updateDiskUsage(path string, add bool) {
-	newDu := float64(newDuInt)
-	lastCheckpointDu := float64(atomic.LoadInt64(&fs.diskUsageCheckpoint))
-	diff := math.Abs(newDu - lastCheckpointDu)
-
-	// If the difference between the checkpointed disk usage and
-	// current one is larger than than 1% of the checkpointed: store it.
-	if (lastCheckpointDu * diskUsageCheckpointPercent / 100.0) < diff {
-		fs.persistDiskUsageFile()
+func (fs *Datastore) checkpointDiskUsage() {
+	select {
+	case fs.checkpointCh <- struct{}{}:
+		// msg sent
+	default:
+		// checkpoint request already pending
 	}
 }
 
-func (fs *Datastore) persistDiskUsageFile() {
-	// Store DiskUsage on clean shutdowns.
-	du, err := fs.DiskUsage()
-	if err != nil {
-		// do not store on errors. just ignore them
-		return
+func (fs *Datastore) checkpointLoop() {
+	timerActive := true
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+	for {
+		select {
+		case _, more := <-fs.checkpointCh:
+			du := atomic.LoadInt64(&fs.diskUsage)
+			fs.dirty = true
+			if !more { // shutting down
+				fs.writeDiskUsageFile(du, true)
+				if fs.dirty {
+					log.Errorf("could not store final value of disk usage to file, future estimates may be inaccurate")
+				}
+				fs.done <- struct{}{}
+				return
+			}
+			// If the difference between the checkpointed disk usage and
+			// current one is larger than than `diskUsageCheckpointPercent`
+			// of the checkpointed: store it.
+			newDu := float64(du)
+			lastCheckpointDu := float64(fs.storedValue.DiskUsage)
+			diff := math.Abs(newDu - lastCheckpointDu)
+			if lastCheckpointDu*diskUsageCheckpointPercent < diff*100.0 {
+				fs.writeDiskUsageFile(du, false)
+			}
+			// Otherwise insure the value will be written to disk after
+			// `diskUsageCheckpointTimeout`
+			if fs.dirty && !timerActive {
+				timer.Reset(diskUsageCheckpointTimeout)
+				timerActive = true
+			}
+		case <-timer.C:
+			timerActive = false
+			if fs.dirty {
+				du := atomic.LoadInt64(&fs.diskUsage)
+				fs.writeDiskUsageFile(du, false)
+			}
+		}
 	}
+}
 
-	atomic.StoreInt64(&fs.diskUsageCheckpoint, int64(du))
-
-	duB := []byte(fmt.Sprintf("%d", du))
+func (fs *Datastore) writeDiskUsageFile(du int64, doSync bool) {
 	tmp, err := ioutil.TempFile(fs.path, "du-")
 	if err != nil {
+		log.Warningf("cound not write disk usage: %v", err)
 		return
 	}
-	if _, err := tmp.Write(duB); err != nil {
+
+	toWrite := fs.storedValue
+	toWrite.DiskUsage = du
+	encoder := json.NewEncoder(tmp)
+	if err := encoder.Encode(&toWrite); err != nil {
+		log.Warningf("cound not write disk usage: %v", err)
 		return
 	}
+
+	if doSync {
+		if err := tmp.Sync(); err != nil {
+			log.Warningf("cound not sync %s: %v", DiskUsageFile, err)
+			return
+		}
+	}
+
 	if err := tmp.Close(); err != nil {
+		log.Warningf("cound not write disk usage: %v", err)
+		return
+	}
+
+	if err := os.Rename(tmp.Name(), filepath.Join(fs.path, DiskUsageFile)); err != nil {
+		log.Warningf("cound not write disk usage: %v", err)
```

Comment:
This should probably be deleting the temporary file on failure (or at least trying to do so).

Reply:
Done.